File: app/src/main/java/com/lifan/app/javasrc/java18/Lambda.java
package com.lifan.app.javasrc.java18;

import java.util.Arrays;
import java.util.function.Consumer;

/**
 * @author by LiFan
 * @date 2019/2/25
 */

class Lambda {
    public static void main(String[] args) {
        new Lambda().setLambda(() -> {
            System.out.println("OK");
        });

        //=============================================================

        Arrays.asList( "a", "b", "d" ).forEach(
                e -> {
                    System.out.println( e );
                    System.out.println( e );
                }
        );

        Arrays.asList( "a", "b", "d" ).forEach(new Consumer<String>() {
            @Override
            public void accept(String s) {
                System.out.println( s );
            }
        });

//        对比而言Kotlin更优雅
//        listOf(1,2,3,4,5).forEach {
//            println(it)
//        }

    }

    @FunctionalInterface
    public interface InterfaceX {
        void doIt();
    }
    public void setLambda(InterfaceX interfaceX){
        interfaceX.doIt();
    }

}

File: src/java_to_kotlin/demo9/Lambda.kt
package java_to_kotlin.demo15

import java.lang.Math.PI

/**
 * @author by LiFan
 * @date 2018/9/10
 */
fun f0() {
    println("A")
}

val f1 = {
    println("A")
}

fun main(args: Array<String>) {
    //运行一个方法
    run(::f0)
    run(f1)

    // 这是一个对象(虽然他是方法)
    val lambda = { println("A") }
    //加()运行它
    { println("A") } ()
    lambda ()
    //使用run 运行它
    run{ println("A") }
    run{ lambda }

    //打印对象(打印方法的入参和返回值)
    println(::f0)
    println(f1)

    //打印匿名函数 { println("A") } (本质上打印入参和返回值)
    println({ println("A") } )
    //运行匿名函数 { println("A") } 打印其返回值kotlin.Unit
    println({ println("A") } ())

    //lambda的申明和定义
    val a1 :()->Double = { PI }
    val a2 :(Int)->Int = { it }
    val a3 :(Int,String)->String = { x,y -> y+x }


}